File: Frontend/lib/tts-client.ts
// TTS client utility following hackathon requirements
interface TTSResponse {
  audioBuffer?: ArrayBuffer
  audioUrl?: string
  error?: string
}

export async function generateAudio(text: string): Promise<TTSResponse> {
  const provider = process.env.TTS_PROVIDER || "azure"

  try {
    switch (provider.toLowerCase()) {
      case "azure":
        return await generateAzureTTS(text)
      case "google":
        return await generateGoogleTTS(text)
      default:
        throw new Error(`Unsupported TTS provider: ${provider}`)
    }
  } catch (error) {
    console.error("TTS generation failed:", error)
    return { error: error instanceof Error ? error.message : "Unknown error" }
  }
}

async function generateAzureTTS(text: string): Promise<TTSResponse> {
  const apiKey = process.env.AZURE_TTS_KEY
  const endpoint = process.env.AZURE_TTS_ENDPOINT

  if (!apiKey || !endpoint) {
    throw new Error("Azure TTS credentials not set")
  }

  const ssml = `
    <speak version="1.0" xmlns="http://www.w3.org/2001/10/synthesis" xml:lang="en-US">
      <voice name="en-US-AriaNeural">
        <prosody rate="medium" pitch="medium">
          ${text}
        </prosody>
      </voice>
    </speak>
  `

  const response = await fetch(`${endpoint}/cognitiveservices/v1`, {
    method: "POST",
    headers: {
      "Ocp-Apim-Subscription-Key": apiKey,
      "Content-Type": "application/ssml+xml",
      "X-Microsoft-OutputFormat": "audio-16khz-128kbitrate-mono-mp3",
    },
    body: ssml,
  })

  if (!response.ok) {
    throw new Error(`Azure TTS API error: ${response.statusText}`)
  }

  const audioBuffer = await response.arrayBuffer()
  return { audioBuffer }
}

async function generateGoogleTTS(text: string): Promise<TTSResponse> {
  const apiKey = process.env.GOOGLE_TTS_API_KEY

  if (!apiKey) {
    throw new Error("Google TTS API key not set")
  }

  const response = await fetch(`https://texttospeech.googleapis.com/v1/text:synthesize?key=${apiKey}`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify({
      input: { text },
      voice: {
        languageCode: "en-US",
        name: "en-US-Wavenet-D",
        ssmlGender: "NEUTRAL",
      },
      audioConfig: {
        audioEncoding: "MP3",
      },
    }),
  })

  if (!response.ok) {
    throw new Error(`Google TTS API error: ${response.statusText}`)
  }

  const data = await response.json()
  const audioBuffer = Buffer.from(data.audioContent, "base64")
  return { audioBuffer: audioBuffer.buffer }
}

File: Backend/src/main/java/com/adobe/hackathon/util/TTSClient.java
package com.adobe.hackathon.util;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.*;
import org.springframework.stereotype.Component;
import org.springframework.web.client.RestTemplate;

import java.io.FileOutputStream;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.Base64;
import java.util.HashMap;
import java.util.Map;

/**
 * TTS Client supporting Azure TTS via environment variables
 * Supports the Adobe India Hackathon 2025 requirements for TTS integration
 */
@Component
public class TTSClient {

    private static final Logger logger = LoggerFactory.getLogger(TTSClient.class);

    @Value("${TTS_PROVIDER:azure}")
    private String ttsProvider;

    @Value("${AZURE_TTS_KEY:}")
    private String azureTTSKey;

    @Value("${AZURE_TTS_ENDPOINT:}")
    private String azureTTSEndpoint;

    @Value("${AZURE_TTS_VOICE:en-US-JennyNeural}")
    private String azureTTSVoice;

    @Value("${AZURE_TTS_FORMAT:mp3}")
    private String azureTTSFormat;

    private final RestTemplate restTemplate;
    private final Path audioDir = Paths.get("./uploads/audio");

    public TTSClient() {
        this.restTemplate = new RestTemplate();
        // Create audio directory if it doesn't exist
        try {
            Files.createDirectories(audioDir);
        } catch (IOException e) {
            logger.warn("Could not create audio directory", e);
        }
    }

    /**
     * Generate audio from text using the configured TTS provider
     */
    public String generateAudio(String text, String jobId) {
        try {
            switch (ttsProvider.toLowerCase()) {
                case "azure":
                    return generateWithAzureTTS(text, jobId);
                default:
                    logger.warn("Unknown TTS provider: {}, using fallback", ttsProvider);
                    return generateFallbackAudio(text, jobId);
            }
        } catch (Exception e) {
            logger.error("Error generating audio with TTS provider: {}", ttsProvider, e);
            return generateFallbackAudio(text, jobId);
        }
    }

    /**
     * Generate audio using Azure TTS
     */
    private String generateWithAzureTTS(String text, String jobId) {
        try {
            String url = azureTTSEndpoint + "/cognitiveservices/v1";
            
            // Prepare SSML (Speech Synthesis Markup Language)
            String ssml = buildSSML(text, azureTTSVoice);

            HttpHeaders headers = new HttpHeaders();
            headers.setContentType(MediaType.valueOf("application/ssml+xml"));
            headers.set("X-Microsoft-OutputFormat", "audio-16khz-128kbitrate-mono-mp3");
            headers.set("Ocp-Apim-Subscription-Key", azureTTSKey);
            headers.set("User-Agent", "AdobeLearnPlatform");

            HttpEntity<String> request = new HttpEntity<>(ssml, headers);
            ResponseEntity<byte[]> response = restTemplate.exchange(url, HttpMethod.POST, request, byte[].class);

            if (response.getStatusCode() == HttpStatus.OK && response.getBody() != null) {
                // Save audio file
                String fileName = "podcast_" + jobId + ".mp3";
                Path audioPath = audioDir.resolve(fileName);
                
                try (FileOutputStream fos = new FileOutputStream(audioPath.toFile())) {
                    fos.write(response.getBody());
                }

                // Return the URL to access the audio file
                return "/api/frontend/audio/" + fileName;
            }

        } catch (Exception e) {
            logger.error("Error generating with Azure TTS", e);
        }
        return generateFallbackAudio(text, jobId);
    }

    /**
     * Build SSML for Azure TTS
     */
    private String buildSSML(String text, String voice) {
        return String.format("""
            <speak version='1.0' xmlns='http://www.w3.org/2001/10/synthesis' xml:lang='en-US'>
                <voice name='%s'>
                    <prosody rate='medium' pitch='medium'>
                        %s
                    </prosody>
                </voice>
            </speak>
            """, voice, escapeXml(text));
    }

    /**
     * Escape XML special characters
     */
    private String escapeXml(String text) {
        return text.replace("&", "&amp;")
                  .replace("<", "&lt;")
                  .replace(">", "&gt;")
                  .replace("\"", "&quot;")
                  .replace("'", "&apos;");
    }

    /**
     * Generate fallback audio when TTS is not available
     */
    private String generateFallbackAudio(String text, String jobId) {
        try {
            // Create a simple text file as fallback
            String fileName = "podcast_" + jobId + ".txt";
            Path audioPath = audioDir.resolve(fileName);
            
            String fallbackContent = "Audio generation not available. Here's the text content:\n\n" + text;
            Files.write(audioPath, fallbackContent.getBytes());

            return "/api/frontend/audio/" + fileName;
        } catch (IOException e) {
            logger.error("Error creating fallback audio", e);
            return null;
        }
    }

    /**
     * Get audio file as byte array
     */
    public byte[] getAudioFile(String fileName) {
        try {
            Path audioPath = audioDir.resolve(fileName);
            if (Files.exists(audioPath)) {
                return Files.readAllBytes(audioPath);
            }
        } catch (IOException e) {
            logger.error("Error reading audio file: {}", fileName, e);
        }
        return null;
    }

    /**
     * Get the current TTS provider configuration
     */
    public String getCurrentProvider() {
        return ttsProvider;
    }

    /**
     * Check if TTS is properly configured
     */
    public boolean isConfigured() {
        switch (ttsProvider.toLowerCase()) {
            case "azure":
                return !azureTTSKey.isEmpty() && !azureTTSEndpoint.isEmpty();
            default:
                return false;
        }
    }

    /**
     * Calculate estimated duration for text (words per minute)
     */
    public int calculateDuration(String text) {
        int wordCount = text.split("\\s+").length;
        int wordsPerMinute = 150; // Average speaking rate
        return Math.max(120, (wordCount * 60) / wordsPerMinute); // Minimum 2 minutes
    }
}
